module type rewrite

hcmh opened the pull request:
Soo,

apparently mxnlinv has higher priority than I thought, and therefore this memory leak (explained in #5 ) does as well.

Here is a first version of the module type (github doesn't like my force-pushed history, that's why I have to make a new pull request).

As far as I can tell, pretty much none of the previous (external) usage of the module has changed. 

Some ugly things in this: Both the module and the device want access to each other's internal implementations. This is, unfortunately, not possible using two files (or at least I couldn't find a way to make it work). So I had to create two access functions to the device, that are not member functions (`const cl_device_id & get_backend_device(const device & d)` and `const cl_context & get_backend_context(const device & d)`). 

One way of achieving this is, I think, putting the module class definition inside of the device class. This would make sense, because the modules really are owned by the device and they live on the device (as evidenced by the `std::unordered_map` used to hold modules). This way, access to modules would necessarily go through the device.
This would break the existing usage of the module type, though. For example, `test/test.backend.kernel` uses modules on their own, which would not be possible anymore.

So, you need to decide which way you want it. I'll implement the same thing for the CUDA side once I know which way to go (and probably a test for multiple kernels in a single module).
And one more thing: the kernels really are pretty low level things, and they get cleaned up properly now. Do you think it still makes sense to give them their own class, too, and changing them from being `typedef`s?

Greetings,
Christian

hcmh committed:
More sane printing of OpenCL build log

Formerly, even an empty module build log would be printed (which still
consists of '\n\0'). Now, we check for that sequence of bytes before printing.
This used to cause lots of blank lines printed on the console on projects
which use a lot of kernels.

Furthermore, the build log is always printed in create_module_from_string. So
before, when the debug paramter in load_from_string was true, the build log
was printed a second time. Now it is always printed, and always printed once.

hcmh committed:
Fix OpenCL memory leak and module rewrite

Formerly, the OpenCL cl_kernel objects were never released. This created a
memory leak for applications which used aura, cleared up all references to it,
and continued running: those applications would continue to use device memory.
This was especially noticable by applications that used aura through Matlabs
mex-interface: Here every invocation of the .mex-file would use up more and
more device memory, until matlab was restarted.

In order to fix this, the cl_kernel objects need to be remembered and cleaned
up using clReleaseKernel(). Since this was not possible in a nice way in the
implementation, a proper module type was created to take care of this.

This new module type abstracts the cl_program type. It takes care of proper
cleanup of both the cl_program and cl_kernel's.

Added feature here is support for multiple kernels in a single module, and
caching of kernels in the module type. (This is stil untested! and only
implemented for OpenCL!)

sschaetz commented:
Looks good. Any thoughts on how to create a module from a string instead of a file? Like if you wan't to pass in the kernel string directly because you created it at runtime or you compiled it into the binary?

hcmh commented:
Because I removed `create_module_from_string`? I think `create_module_from_file` should also just be for internal use, and `load_from_file/string` on the `device` object should be used for public access. 

And `create_module_from_string` lives on as the constructor for the `module` type. If there were a good way of distinguishing between a file name and a module string, I would have removed `create_module_from_file` as well and added it as another contructor....

So I guess if you really need it, you coul still just call the constructor directly. But I see the ugliness in breaking the symmetry between `create_module_from_file/string`

sschaetz commented:
You're right. The one constructor of module is what confused me. But the device.load_from\* is what should be called. I'll merge, would be great if you could also implement the CUDA version if you find time. Thanks for the contribution.
